ambry-store/src/main/java/com/github/ambry/store/BlobStoreStats.java

```diff
@@ -706,46 +766,132 @@ private void processNewPut(ScanResults results, IndexValue putValue) {
       if (originalPutValue.getExpiresAtMs() != Utils.Infinite_Time) {
         // make appropriate updates to avoid double counting
-        handleContainerBucketUpdate(results, originalPutValue, originalPutValue.getExpiresAtMs(), ADD);
-        handleLogSegmentBucketUpdate(results, originalPutValue, originalPutValue.getExpiresAtMs(), ADD);
+        handleLogSegmentExpiredBucketUpdate(results, originalPutValue, originalPutValue.getExpiresAtMs(),
+            operator * ADD);
+      } else if (deleteValue.isTtlUpdate()) {
+        // This blob has a PUT and TTL_UPDATE, because of DELETE, now TTL_UPDATE is not valid any more.
+        try {
+          IndexValue ttlUpdateValue =
+              index.findKey(key, new FileSpan(originalPutValue.getOffset(), deleteValue.getOffset()),
+                  EnumSet.of(PersistentIndex.IndexEntryType.TTL_UPDATE));
+          if (ttlUpdateValue != null) {
+            handleLogSegmentDeletedBucketUpdate(results, ttlUpdateValue, operationTimeInMs, operator * SUBTRACT);
+          }
+        } catch (StoreException e) {
+          logger.error(
+              "Failed to find TTL_UPDATE IndexValue for " + key + " when processing new DELETE: " + deleteValue, e);
+        }
       }
     }
   }
 
   /**
    * Helper function to process new UNDELETE entries and make appropriate updates to the given {@link ScanResults}.
    * @param results the {@link ScanResults} to apply the updates to
+   * @param key the {@link StoreKey} of the new UNDELETE
    * @param undeleteValue the {@link IndexValue} of the new UNDELETE
    * @param originalPutValue the {@link IndexValue} of the original PUT
+   * @param previousValue the {@link IndexValue} comes prior to the new UNDELETE
    */
-  private void processNewUndelete(ScanResults results, IndexValue undeleteValue, IndexValue originalPutValue) {
-    // TODO: future work to support online updates
+  private void processNewUndelete(ScanResults results, StoreKey key, IndexValue undeleteValue,
+      IndexValue originalPutValue, IndexValue previousValue) {
+    // First deal with container bucket
+    if (!previousValue.isDelete()) {
+      // This previous value is not DELETE, current UNDELETE would not change container bucket.
+      // There are several possibility when the previous value is not DELETE.
+      // 1. PUT UNDELETE: UNDELETE doesn't change any bucket
+      // 2. PUT TTL_UPDATE UNDELETE: UNDELETE doesn't change any bucket
+      // 3. PUT DELETE UNDELETE UNDELETE: the last UNDELETE doesn't change any bucket
+      // UNDELETE would resurrect a deleted IndexValue, if the previous value is not DELETE, then the IndexValue doesn't
+      // need to be resurrected.
+    } else {
+      handleContainerBucketUpdate(results, originalPutValue, undeleteValue.getOperationTimeInMs(), ADD);
+      handleContainerBucketUpdate(results, originalPutValue, originalPutValue.getExpiresAtMs(), SUBTRACT);
```

Comment:
I understand this part and just want to confirm again about UNDELETE. The UNDELETE operation always checks if original PUT has expired. So, if we reach here, it means original PUT hasn't expired and should be subtracted, correct?

I am asking this because I am thinking about a case where original PUT expires before DELETE and there is an UNDELETE in the end. In this case, original PUT is subtracted twice but I think this case would never happen.

Reply:
Your first statement is correct, and there is no chance when the PUT is expired before DELETE and UNDELETE would succeeds. If we have PUT and DELETE for the blob, but PUT is already expired, then UNDELETE will fail since it can't undelete an expired PUT.